### PR TITLE
Adds Glossaries functionality 

### PR DESCRIPTION
**Pull request by ortegacmanuel**
This PR adds the possibility to create glossaries in the inline_forms back-end and check words as part of one of more glossaries. In the front end users can explore the glossaries and download them as a PDF document. Technically this PD adds:

- New Glossary model that `has_and_belongs_to_many` words
- Database migrations for the Glossary model and new table for its relation with the words table
- New views in the frontend and its methods en the FrontendController
- The pdfkit gem
- Views and logic for pdf generation

glossary_frontend
glossary_backend

### CODE REVIEW

**Comment by acesuares** [comment]
Seems some file should not be checked in!!!!!

File: app/controllers/frontends_controller.rb [hunk 1]
```diff
@@ -27,6 +27,40 @@ def palabra
     end
   end
 
+  def glosario
+    if params[:glossary].blank?
+      redirect_to '/'
+    else
+      @search_glossary = params[:glossary].squish.gsub(CHARACTER_REGEX,'')
```
Comment (acesuares) [inline app/controllers/frontends_controller.rb]: why clean the search term? it won't find it when it's bogus.. and mysql injection is already dealt with... the way you do it, word and word_ and w_?ord all match...

File: app/controllers/frontends_controller.rb [hunk 2]
```diff
@@ -27,6 +27,40 @@ def palabra
     end
   end
 
+  def glosario
+    if params[:glossary].blank?
```
Comment (acesuares) [inline app/controllers/frontends_controller.rb]: search for blank returns nil, too so...

File: app/controllers/frontends_controller.rb [hunk 3]
```diff
@@ -27,6 +27,40 @@ def palabra
...
+    else
+      @search_glossary = params[:glossary].squish.gsub(CHARACTER_REGEX,'')
+      @glossary = Glossary.find_by_name(@search_glossary)
+      if @glossary.nil?
+        redirect_to '/'
+      end
+    end
+    respond_to do |format|
+      format.html {
+      }
```
Comment (acesuares) [inline app/controllers/frontends_controller.rb]: add at least some html

File: app/controllers/frontends_controller.rb [hunk 4]
```diff
@@ -27,6 +27,40 @@ def palabra
...
+      format.pdf {
+        html = render_to_string(layout: 'pdf')
+        filename = "#{Time.now.to_s(:db)[0..9]}_#{@glossary.name.strip}.pdf"
+        if Rails.env.production?
+         # generate the pdf
+         html = html.gsub(/\/assets\//, 'https://sazon.id-arte.net/assets/')
+         html = html.gsub(/\/uploads\//, 'https://sazon.id-arte.net/uploads/')
+         pdf = PDFKit.new(html).to_pdf
```
Comment (acesuares) [inline app/controllers/frontends_controller.rb]: these last two lines are generic, move them out of the format

File: app/helpers/frontends_helper.rb [hunk 5]
```diff
@@ -10,4 +10,16 @@ def link_to_word(word, full_url=false)
...
+
+  def link_to_glossary(glossary, full_url=false)
+    if full_url
+      if Rails.env.development?
+        link_to glossary.name, "http://localhost:3000/glosario/#{glossary.name}" unless glossary.nil?
+      else
+        link_to glossary.name, "https://www.papiamentu.info/glosario/#{glossary.name}" unless glossary.nil?
+      end
+    else
+      link_to glossary.name, "/glosario/#{glossary.name}" unless glossary.nil?
```
Comment (acesuares) [inline app/helpers/frontends_helper.rb]: what happens if glossary.nil?

File: app/models/glossary.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    ]
+  end
+
+
+  def self.not_accessible_through_html?
+    false
+  end
+
+  def self.order_by_clause
+    "name"
```
Comment (acesuares) [inline app/models/glossary.rb]: make that 'nil'.. we need to remove that thing!

File: app/models/word.rb [hunk 7]
```diff
@@ -67,6 +68,8 @@ def inline_forms_attribute_list
       [ :tr_pap_cw , "tr_pap_cw", :text_field ],
       [ :tr_pap_aw , "tr_pap_aw", :text_field ],
       [ :buki_di_oro_text, '', :info ],
+      [ :header_otro, '', :header ],
```
Comment (acesuares) [inline app/models/word.rb]: should be called header_glosario

File: app/views/frontends/glosario.html.erb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  <%= render "/frontends/papiamentu_header" %>
+  <div class="row" id="left-side">
+    <div class="columns small-12 medium-6 large-8 topper  bordered">
+      <h2>
+        <%= @glossary.title %>
+      </h2>
+
+      <div class="pap_description bordered">
+        <%= content_tag :p, @glossary.description %>
+        <% alpha = @glossary.words.order("name ASC").group_by{|u| u.name[0]}
```
Comment (acesuares) [inline app/views/frontends/glosario.html.erb]: nice!

File: app/views/frontends/glosario.html.erb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+    <div class="columns small-12 medium-6 large-4 topper bordered" id="right-side">
+      <h3>Spònser</h3>
+      <p>A logra traha e glosario aki danki na ayudo i finansamentu di:</p>
+      <ul>
+        <li>spònser 1</li>
+        <li>spònser 2</li>
+        <li>spònser 3</li>
+      </ul>
+      <%= link_to('Baha dokumento PDF', "/glosario/#{@glossary.name.strip}.pdf") %>
```
Comment (acesuares) [inline app/views/frontends/glosario.html.erb]: why strip?

File: config/initializers/assets.rb [hunk 10]
```diff
@@ -9,4 +9,4 @@
 # Precompile additional assets.
 # application.js, application.css, and all non-JS/CSS in app/assets folder are already added.
 # Rails.application.config.assets.precompile += %w( search.js )
-  Rails.application.config.assets.precompile += %w( devise.css )
+  Rails.application.config.assets.precompile += %w( devise.css pdf.css)
```
Comment (acesuares) [inline config/initializers/assets.rb]: should be named glosario_pdf.css ?

File: config/routes.rb [hunk 11]
```diff
@@ -1,5 +1,9 @@
 Rails.application.routes.draw do
 
+  resources :glossaries do
```
Comment (acesuares) [inline config/routes.rb]: if this is in MODELS then don't need to add route

File: pdfs/2018-02-01_Kuido.pdf.html [hunk 12]
```diff
@@ -0,0 +1,281 @@
+<!DOCTYPE html>
```
Comment (acesuares) [inline pdfs/2018-02-01_Kuido.pdf.html]: should not be in git

File: pdfs/2018-02-01_Kòmpiuter .pdf.html [hunk 13]
```diff
@@ -0,0 +1,88 @@
+<!DOCTYPE html>
```
Comment (acesuares) [inline pdfs/2018-02-01_Kòmpiuter .pdf.html]: shoud not be in git

File: app/views/frontends/glosario.html.erb (hunk 8 above)
Comment (ortegacmanuel) [reply app/views/frontends/glosario.html.erb]: :)

File: app/controllers/frontends_controller.rb (hunk 2 above)
Comment (ortegacmanuel) [reply app/controllers/frontends_controller.rb]: Ok!

File: app/controllers/frontends_controller.rb (hunk 1 above)
Comment (ortegacmanuel) [reply app/controllers/frontends_controller.rb]: I did a copy/paste from your code in CAV, now I understand why this have no sense here..already removed

File: app/controllers/frontends_controller.rb (hunk 3 above)
Comment (ortegacmanuel) [reply app/controllers/frontends_controller.rb]: The html is in the view..which name is glosario.html.erb

File: app/helpers/frontends_helper.rb (hunk 5 above)
Comment (ortegacmanuel) [reply app/helpers/frontends_helper.rb]: If glosarry is nil, no link is returned

**Comment by ortegacmanuel** [comment]
@acesuares All changes done!

File: app/views/layouts/glosario.pdf.erb [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+  <body>
+    <div id="glossary_pdf_header">
+      <%= image_tag 'papiamentu_info_logo.png' %>
+    </div>
+    <div id="sazon_pdf_outer_container">
+      <%= yield %>
+    </div>
+    <div id="glossary_pdf_footer">
+      Teksto i Diseño : Copyright © 2011-2017 FMS / Papiamentu.info, Software: © 2011-2017 Suares &amp; Co, Licensed under the MIT license.
```
Comment (acesuares) [inline app/views/layouts/glosario.pdf.erb]: FMS should not be there! We need to fix that. We need to put the copyright on Suares & Co or on one of the foundations, and then license the glossary under cc.

**Pull request merged**